tests/ssh_tests.py

```diff
@@ -83,13 +85,23 @@ def trying_to_use_ssh_shell_after_exit_results_in_error():
     assert_raises(Exception, lambda: shell.run(["true"]))
 
 
-def _create_shell_with_wrong_port():
+@istest
+def passing_a_custom_socket_like_a_pre_connected_one():
+    sock = socket.socket(socket.AddressFamily.AF_INET, socket.SOCK_STREAM)
+    retry_on_signal(lambda: sock.connect((HOSTNAME, PORT)))
+
+    with _create_shell_with_wrong_port(sock=sock) as shell:
+        shell.run(["true"])
+
+
+def _create_shell_with_wrong_port(sock=None):
     return spur.SshShell(
-        username="bob",
-        password="password1",
-        hostname="localhost",
+        username=USERNAME,
+        password=PASSWORD,
+        hostname=HOSTNAME,
         port=54321,
         missing_host_key=spur.ssh.MissingHostKey.accept,
+        sock=sock
```

Comment:
Actually... in the second part, where you provide a proxy command, i realized that my command has to have the destination hostname/IP.  So to simplify that it would be great if it could reuse the hostname given to SshShell.

So i was planning on having a SockProvider class that would have a get_sock(target_hostname) method that a spur.SshProxyCommand would implement and be passed as a sock_provider=  rather that sock=

The whole goal for this is to keep your library "high level" and "completely" hide paramiko.  Passing in kwargs that would be sent to paramiko would require a user to know paramiko.  I'm not against it though.

Still in for the kwargs?